build.gradle

```diff
@@ -1099,15 +1103,17 @@ project(':core') {
       implementation libs.dropwizardMetrics
       exclude module: 'slf4j-log4j12'
       exclude module: 'log4j'
-      // Both Kafka and Zookeeper use slf4j. ZooKeeper moved from log4j to logback in v3.8.0, but Kafka relies on reload4j.
+      // Both Kafka and Zookeeper use slf4j. ZooKeeper moved from log4j to logback in v3.8.0.
       // We are removing Zookeeper's dependency on logback so we have a singular logging backend.
       exclude module: 'logback-classic'
       exclude module: 'logback-core'
     }
     // ZooKeeperMain depends on commons-cli but declares the dependency as `provided`
     implementation libs.commonsCli
-
-    compileOnly libs.reload4j
+    implementation libs.log4j2Core
+    implementation libs.log4j2Api
+    implementation libs.log4j1Bridge2Api
+    implementation libs.jacksonDatabindYaml
```

Comment:
How come these are `implementation` while `slf4jLog4j` is `testImplementation`?

Reply:
[`Log4jController`](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/utils/Log4jController.scala#L21) needs it to configure the log level at runtime

Reply:
`jacksonDatabindYaml` is required to parse yaml config - see https://logging.apache.org/log4j/2.x/manual/installation.html#impl-core-config

however, it seems we can do a bit cleanup for `jacksonDatabindYaml` as not all modules need to add explicit reference

Reply:
We should perhaps bite the bullet and just make the slf4jLog4j choice for the server. After all, unless you choose this, some functionality won't work.

Reply:
IMHO the `kafka_<scala_version>` artifact should not have these dependencies, these dependencies should be added **only** to the binary Kafka distribution. Otherwise Kafka will leak the Log4j Core dependencies to its consumers, similarly to what was happening with Zookeeper (see [ZOOKEEPER-4820](https://issues.apache.org/jira/browse/ZOOKEEPER-4820)).

To add dependencies only to the binary distribution, you could use something similar to apache/eventmesh#4719, i.e. a separate `distOnly` Gradle configuration.

`Log4jController` can be rewritten to use Log4j Core if present or a no-op implementation otherwise, so the Log4j Core can be declared as an [optional Gradle dependency](https://blog.gradle.org/optional-dependencies).

**Note**: I am working on a `org.apache.logging:logging-admin` artifact (see [`ppkarwasz/logging-admin` repo](https://github.com/ppkarwasz/logging-admin)) that would provide the same functionality as `Log4jController`, but in a logging implementation independent way. Unfortunately I have a long TODO list before I can publish it, so probably it won't be ready for Kafka 4.

Reply:
`Log4jController` is used exclusively by the server, so exposing its dependencies should be acceptable.

Reply:
In Kafka 3.9.0, the dependency on `ch.qos.reload4j:reload4j` is declared `compileOnly`:

https://github.com/apache/kafka/blob/84caaa6e9da06435411510a81fa321d4f99c351f/build.gradle#L970

The dependency is added to the binary distribution in a separate task:

https://github.com/apache/kafka/blob/84caaa6e9da06435411510a81fa321d4f99c351f/build.gradle#L1027-L1037

These dependencies should probably be handled in a similar way.

Reply:
**Note**: No modern library uses Log4j 1 in code (they use JCL, SLF4J or Log4j API), so my guess is that `libs.log4j1Bridge2Api` could be dropped entirely.

Reply:
> These dependencies should probably be handled in a similar way.

Yes, we can declare them as compileOnly and then add them to the distribution. However, the flexibility of replacing the SLF4J provider at runtime may break the functionality of Log4jController (similar to [Juma's comment](https://github.com/apache/kafka/pull/17373#discussion_r1894553658)). KIP-1064 is attempting to use SLF4J2's system variable to choose the provider more effectively.

> No modern library uses Log4j 1 in code (they use JCL, SLF4J or Log4j API), so my guess is that libs.log4j1Bridge2Api could be dropped entirely.

Pardon me, in the #18290 we decide to allow users to use log4j.properties - so we still need `log4j-1.2-api`, right?

Reply:
> Yes, we can declare them as compileOnly and then add them to the distribution. However, the flexibility of replacing the SLF4J provider at runtime may break the functionality of Log4jController (similar to [Juma's comment](https://github.com/apache/kafka/pull/17373#discussion_r1894553658)). KIP-1064 is attempting to use SLF4J2's system variable to choose the provider more effectively.

As far as I can tell, this is the way `Log4jController` worked in Kafka 3.x: if the optional `ch.qos.reload4j` dependency was absent, the class didn't work.

> Pardon me, in the #18290 we decide to allow users to use log4j.properties - so we still need `log4j-1.2-api`, right?

Sorry, my mistake.

Reply:
> As far as I can tell, this is the way Log4jController worked in Kafka 3.x: if the optional ch.qos.reload4j dependency was absent, the class didn't work.

Yes, you're right. Perhaps we should consider offering similar functionality for other popular SLF4J providers, such as Logback and JUL. @ijuma WDYT?

Reply:
@ppkarwasz It is true that we previously tried hard to avoid making the logging choice for the Maven artifact while making it for the distributed binaries. However, this is brittle and only worked partially. When I tried to fix it in #12148, it caused problems and it was partially reverted (#16260, #16559).

Reply:
> 1. Your logging admin library.

I pushed the draft to Apache Logging ([`apache/logging-admin`](https://github.com/apache/logging-admin)) and I'll start to actively work on it. Probably you can expect a release by end of January/February.

In the meantime I can make a PR for Kafka, so that `Log4jController` fails softly if Log4j Core is not present.

> 2. slf4j2 makes it possible to choose the logging library dynamically instead of via classpath tricks.

Note that choosing the SLF4J implementation does not really tell you which logging implementation is being used: except Logback, all the other SLF4J implementation are bridges between logging APIs. If you use `slf4j-jdk14` you don't know which JUL implementation is being used and if you use `log4j-slf4j2-impl` you don't know which Log4j API implementation is being used.

Reply:
> In the meantime I can make a PR for Kafka, so that Log4jController fails softly if Log4j Core is not present.

Yes, it would be great to display accurate error messages!